ocp4crd/oauth/oauth.go

```diff
@@ -0,0 +1,135 @@
+	Name          string `yaml:"name"`
+	Challenge     bool   `yaml:"challenge"`
+	Login         bool   `yaml:"login"`
+	MappingMethod string `yaml:"mappingMethod"`
+	Type          string `yaml:"type"`
+	GitHub        struct {
+		HostName string `yaml:"hostname"`
+		CA       struct {
+			Name string `yaml:"name"`
+		} `yaml:"ca"`
+		ClientID     string `yaml:"clientID"`
+		ClientSecret struct {
+			Name string `yaml:"name"`
+		} `yaml:"clientSecret"`
+		Organizations []string `yaml:"organizations"`
+		Teams         []string `yaml:"teams"`
+	} `yaml:"github"`
+}
+
+type identityProviderHTPasswd struct {
+	Name          string `yaml:"name"`
+	Challenge     bool   `yaml:"challenge"`
+	Login         bool   `yaml:"login"`
+	MappingMethod string `yaml:"mappingMethod"`
+	Type          string `yaml:"type"`
+	HTPasswd      struct {
+		FileData struct {
+			Name string `yaml:"name"`
+		} `yaml:"fileData"`
+	} `yaml:"htpasswd"`
+}
+
+// Shared CRD part, present in all types of OAuth CRDs
+type v4OAuthCRD struct {
+	APIVersion string `yaml:"apiVersion"`
+	Kind       string `yaml:"kind"`
+	MetaData   struct {
+		Name string `yaml:"name"`
+	} `yaml:"metadata"`
+	Spec struct {
+		IdentityProviders []interface{} `yaml:"identityProviders"`
+	} `yaml:"spec"`
+}
+
+// Generate converts OCPv3 OAuth to OCPv4 OAuth Custom Resources
+func Generate(masterconfig configv1.MasterConfig) error {
+	var auth = masterconfig.OAuthConfig
+	var err error
+
+	var crd v4OAuthCRD
+	crd.APIVersion = "config.openshift.io/v1"
+	crd.Kind = "OAuth"
+	crd.MetaData.Name = "cluster"
+
+	serializer := json.NewYAMLSerializer(json.DefaultMetaFactory, scheme.Scheme, scheme.Scheme)
+	for _, p := range auth.IdentityProviders {
+		p.Provider.Object, _, err = serializer.Decode(p.Provider.Raw, nil, nil)
+
+		switch kind := p.Provider.Object.GetObjectKind().GroupVersionKind().Kind; kind {
+		case "HTPasswdPasswordIdentityProvider":
+			var idP identityProviderHTPasswd
+			var htpasswd configv1.HTPasswdPasswordIdentityProvider
+			_, _, _ = serializer.Decode(p.Provider.Raw, nil, &htpasswd)
+
+			idP.Type = "HTPasswd"
+			idP.Challenge = p.UseAsChallenger
+			idP.Login = p.UseAsLogin
+			idP.MappingMethod = p.MappingMethod
+			idP.HTPasswd.FileData.Name = htpasswd.File
+
+			crd.Spec.IdentityProviders = append(crd.Spec.IdentityProviders, idP)
+		case "GitHubIdentityProvider":
+			var idP identityProviderGitHub
+			var github configv1.GitHubIdentityProvider
+			_, _, _ = serializer.Decode(p.Provider.Raw, nil, &github)
+
+			idP.Type = "GitHub"
+			idP.Name = p.Name
+			idP.Challenge = p.UseAsChallenger
+			idP.Login = p.UseAsLogin
+			idP.MappingMethod = p.MappingMethod
+			idP.GitHub.HostName = github.Hostname
+			idP.GitHub.CA.Name = github.CA
+			idP.GitHub.ClientID = github.ClientID
+			idP.GitHub.Organizations = github.Organizations
+			idP.GitHub.Teams = github.Teams
+			// TODO: Learn how to handle secrets
+			idP.GitHub.ClientSecret.Name = github.ClientSecret.Value
```

Comment:
Secrets can be in different places: ENV var, configmap, secrets, plaintext. We have to learn how to handle those.